vm/src/dictdatatype.rs

```diff
@@ -573,13 +606,35 @@ impl<T: Clone> Dict<T> {
         Ok(removed)
     }
 
-    pub fn pop_back(&self) -> Option<(PyObjectRef, T)> {
-        let mut inner = self.write();
-        inner.entries.pop().map(|entry| {
-            inner.used -= 1;
-            inner.indices[entry.index] = IndexEntry::DUMMY;
-            (entry.key, entry.value)
-        })
+    pub fn pop_back(&self, vm: &VirtualMachine) -> Option<(PyObjectRef, T)> {
+        let inner = self.read();
+        let mut idx: i64 = inner.nentries as i64 - 1;
+
+        while idx >= 0 && inner.entries.get(idx as usize).unwrap().is_none() {
+            idx -= 1;
+        }
+
+        if idx < 0 {
+            None
+        } else {
+            let idx = idx as usize;
+            let entry = inner.entries.get(idx).unwrap().as_ref().unwrap().clone();
+
+            let removed_key = entry.key;
+            let removed_item = entry.value;
+            if let Ok(lookup) = self.lookup(vm, &removed_key, entry.hash, Some(inner)) {
+                match self.pop_inner(lookup) {
+                    Ok(_) => {
+                        let mut inner = self.write();
+                        inner.nentries = idx as usize;
+                        Some((removed_key, removed_item))
+                    }
+                    Err(_) => None,
+                }
+            } else {
+                None
+            }
+        }
```

Comment:
gramatical suggestion:
```suggestion
        let mut idx = inner.nentries.checked_sub(1)?;

        while inner.entries.get(idx as usize).unwrap().is_none() {
            idx = idx.checked_sub(1)?;
        }

        let entry = inner.entries.get(idx).unwrap().as_ref().unwrap().clone();

        let removed_key = entry.key;
        let removed_item = entry.value;
        let lookup = self.lookup(vm, &removed_key, entry.hash, Some(inner)).ok()?;
        self.pop_inner(lookup).ok()?;
        self.write().nentries = idx as usize;
        Some((removed_key, removed_item))
```

But this implementation looks a bit weird for me.
1. do we need to copy the entry?
2. do we need to lookup to pop_back?

Reply:
1. only removed_key and removed_item should be cloned, but i copied entry to use clone once and I thought it is more clean to see the code. if it is better to clone only the two things, i'll fix that
2. It is totally redundant. i'll fix that

Reply:
For the clones, I don't think we need any clone for this operation in best solution.
(though I didn't looked in it that enough yet)

Reply:
I think the entry should be moved out rather than being cloned. Please check `std::mem::take` part of other comment.